### DISCUSSION TITLE
Order of focus for group of controls seems to be indeterministic, need preliminary help

### DISCUSSION THREAD

**Discussion by dwiyatci** [react-hook-form/react-hook-form#3708] [General] [answered]
Hi, in my project I noticed that the order of focus for group of controls doesn't consistently follow the order of their occurrences in the DOM – even though I've used `<Controller />` for all my controls. It happened during focus-on-error and focus-on-insert item using `useFieldArray`.  

I searched the issues threads, and it was mentioned somewhere that the focus order is basically the `register()` order of the inputs/controls. Now, since I'm not really sure that I can reproduce it on CSB, I would like to know first the part of lib code that does it. Could you point me out the relevant files/lines for this "registration process" of the controls that determines their focus order in the source code? ;)

Thank you.

**Comment by bluebill1049** [comment, marked as answer]
- Controller register at useEffect
- Native input gets registered right away

that's the difference.

**Reply by dwiyatci** [reply]
I see. And is there a variable that I can debug/console.log the order somehow? As said, I don't use any native input, but instead I only use `<Controller>`s.

**Reply by bluebill1049** [reply]
no, we don't. perhaps next major version we can look at let user order the focus order, but not sure how nice an API will that be. it's a tricky one. Order is always hard because it's not mattered by the display position of the inputs rather when it's get rendered and gets registered in hook form.

**Reply by dwiyatci** [reply]
Ahah, I understand better the tricky-ness now. Yeah, actually that's exactly the least I'm expecting 😁: it's ordered by the order they specified in the code / appear in the DOM. But ok, crossing fingers for the next major version then. Thanks so much for the effort, Bill 🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️

**Reply by bluebill1049** [reply]
Thanks for understanding @dwiyatci i will see what we can come up with next version.
